### ISSUE TITLE
Like operator edge cases 

### ISSUE THREAD

**Issue by cristianbuse** [WaynePhillipsEA/twinbasic#209]
Some of the following tests are failing in tB:  
```VBA
Module TestLikeBinary
    Option Compare Binary
    Public Sub TestLike()
        Debug.Assert ("Xi" Like "[S-\]i") = True
        Debug.Assert (":i" Like "[S-\]i") = False
        Debug.Assert ("\i" Like "[S-\]i") = True
        Debug.Assert (":c8:""fycMFP^jB.b|=|D])cn\x96A,""G{U]brp.G85Vg`{`g10" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False
        Debug.Assert (":48:|VziMeY^PPv{eVkBcK""6$ E(?l8{dm^`*jTnMN:D80" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False
        Debug.Assert (":(8:wOtpMJ3^(!7 %qgW7b3T~;]P(!`#si:s`_w/+2el_:R00" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False

        ' Check error cases
        On Error Resume Next
        Debug.Assert ("(mq8@rqGX2y8!>YC]f17|0^v8ihD.+9q_LU6!8z%qd9X.J" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("(mWP|_aO}q0_:reE)Z*_=(TWox4.dm>piDWQMWi?8r8Q3_.?" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("(m`~DBKdrL+27~!?c;>.>Mw#h[iFh0x>&<]/s/Ix}Hhc3/^}iUh6m.L" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear

        On Error GoTo 0
    End Sub
End Module
Module TestLikeText
    Option Compare Text
    Public Sub TestLike()
        Debug.Assert (":c8:""fycMFP^jB.b|=|D])cn\x96A,""G{U]brp.G85Vg`{`g10" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False
        Debug.Assert (":48:|VziMeY^PPv{eVkBcK""6$ E(?l8{dm^`*jTnMN:D80" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False
        Debug.Assert (":(8:wOtpMJ3^(!7 %qgW7b3T~;]P(!`#si:s`_w/+2el_:R00" Like ":?8:[!1-j'-+]?[!)-\!-dQ-s8-B][!:-N]M?[!]^**???*#0") = False

        ' Check error cases
        On Error Resume Next
        Debug.Assert ("(mq8@rqGX2y8!>YC]f17|0^v8ihD.+9q_LU6!8z%qd9X.J" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("(mWP|_aO}q0_:reE)Z*_=(TWox4.dm>piDWQMWi?8r8Q3_.?" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("(m`~DBKdrL+27~!?c;>.>Mw#h[iFh0x>&<]/s/Ix}Hhc3/^}iUh6m.L" Like "(m*?[!<-YK-K]****?[N-o]#[4-i[-t].[^-z!-z]") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("Xi" Like "[S-\]i") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert (":i" Like "[S-\]i") = True
        Debug.Assert Err.Number <> 0
        Err.Clear
        Debug.Assert ("\i" Like "[S-\]i") = True
        Debug.Assert Err.Number <> 0
        Err.Clear

        On Error GoTo 0
    End Sub
End Module
```

It looks quite strange because if I add the ```Err.Number``` as an expression in the Watch window and step through code line by line the error number only changes on the first ```Debug.Assert Err.Number <> 0``` line.

In order to make testing easier I wrote a test generator in VBA. The immediate advantage is that all tests are checked in VBA so I know the asserts must be correct (unless VBA is flawed as well in some of them).

[Here](https://github.com/cristianbuse/twinBASIC-Like-Test-Generator/blob/master/demo%20tests/10000%20tests%20-%20tB%20Like%20operator.txt?raw=true) are 10000 tests generated with [twinBASIC-Like-Test-Generator](https://github.com/cristianbuse/twinBASIC-Like-Test-Generator).

**Comment by WaynePhillipsEA** [comment]
Just to clarify, are you seeing the first three asserts in TestLikeBinary fail, as they don't fail here?  Same goes for the final 3 cases in TestLikeText?

**Comment by WaynePhillipsEA** [comment]
QuirkAlert:  VBA accepts a pattern of "[K-K]" but throws an error on "*[K-K]".  It appears that it disallows the range start and end to be the same character only when the asterisk appears earlier in the pattern.

**Comment by Kr00l** [comment]
> QuirkAlert: VBA accepts a pattern of "[K-K]" but throws an error on "*[K-K]". It appears that it disallows the range start and end to be the same character only when the asterisk appears earlier in the pattern.

If a pattern raises an Error in VBx, but is "ok" in tB. Then it would not harm? And you avoid a quirk with potential regressions.

**Comment by cristianbuse** [comment]
> Just to clarify, are you seeing the first three asserts in TestLikeBinary fail, as they don't fail here? Same goes for the final 3 cases in TestLikeText?

No, the first 3 asserts in TestLikeBinary do no fail. Sorry, I simply copy-pasted the generated tests.
However, the remaining 6 tests are all failing in TestLikeBinary.
Meanwhile in TestLikeText only tests 3, 4, 5 and 6 are failing.

**Comment by cristianbuse** [comment]
> QuirkAlert: VBA accepts a pattern of "[K-K]" but throws an error on "*[K-K]". It appears that it disallows the range start and end to be the same character only when the asterisk appears earlier in the pattern.

Is this a VBA-only quirk or does it happen in VB6 also?
It seems that ```"*#?[K-K]"``` or ```"*#?[a-tK-K]"``` are also not allowed in VBA so any preceding asterix dissalows the 1-elem range regardless if there are other ranges in the charlist or any other characters in between.

**Comment by Kr00l** [comment]
@cristianbuse both are error-ing in VB6 also.
![image](https://user-images.githubusercontent.com/33756666/121478349-39c8c300-c9c9-11eb-97b8-160f4655e15c.png)

**Comment by cristianbuse** [comment]
@Kr00l Thanks, good to know!

@WaynePhillipsEA  I've updated the test generator (and the 10000 tests) so that the quirk is completely excluded. I figured you don't want to debug 100 failed tests and find out that the same quirk is causing them all. As you find more quirks, I am happy to update the generator.

**Comment by WaynePhillipsEA** [comment]
@cristianbuse I've already copied the quirk over into tB, as I see no harm in replicating it.  No worries though, I've got the original 10000 tests from earlier.  Thanks!

**Comment by WaynePhillipsEA** [comment]
QuirkAlert:  ("abc" Like "*b[!-x]")    gives FALSE in VB, when it should give TRUE.   
Note that ("abc" Like "ab[!-x]") gives the correct answer.

**Comment by WaynePhillipsEA** [comment]
Ahhh.. I see what VB is doing wrong with that quirk.  It's interpretting the '-x' wrongly as the range Chr(0) - 'x', and so excluding all chars up to 'x'

**Comment by WaynePhillipsEA** [comment]
QuirkAlert:  ("aa" Like "*[[-{]") gives FALSE in VB, when it should give TRUE.
Note that ("aa" Like "a[[-{]") gives the correct answer

With this one, VB ignores the '[' to '{' range, and instead only matches to '['.

**Comment by WaynePhillipsEA** [comment]
@cristianbuse Thanks very much for sending me down this rabbit hole!   Remind me to return the favour sometime 👌

J/k...  there were some really nice edge cases in there, so thanks a lot.    Nearly all of the expected test cases are now working in v0.10.1444 (from both the first and second 100,000 dump).  However, many of the error-tests are not throwing errors.  It's a difficult balance to try to match VB here, especially as VB hard-crashes on many of the error-tests.  I will revisit this when I have more time.

For the curious, here are the only tests that are still failing (apart from the error-tests as mentioned above):
 ```
Module TestLikeBinary06

    Option Compare Binary
    Public Sub RunTests()

        Debug.Assert ("/&hpk.0I94]" Like "*[!#-&/-Z)--[-p]#?##?[]") = False
        Debug.Assert ("#OJNr`/x-Ch<qbC35""" Like "*[!#-&/-Z)--[-p]#?##?[]") = False
        Debug.Assert ("q!DcZcd!6j65\" Like "*[!#-&/-Z)--[-p]#?##?[]") = True

        Debug.Assert ("5<@DM_LrL^n Zj!$yk;Qp8Vmf\GuHYhChYyOVhyHU2QPsTDi\q|6;o2^*v{b5'/0oGN!4f$tf vT2P$J^>k^+dEaje]@mrkI)QM" Like "#<@?*m*[?-W-O-bE-K]#Q*#?[]*N**?") = True
        Debug.Assert ("8<@;)Z-6-?j;Bo|ncm26rK2]eT9Qx:X D?Fxh{Po`Wk3}=U`.~_56l4h9&_%eo%~NMFma+'|w>on4+X?\CX>" Like "#<@?*m*[?-W-O-bE-K]#Q*#?[]*N**?") = True
        Debug.Assert ("7<@XpEmK]9QAfk4]uI,X7*9|}Kg48PUlI-+)p9},4!C4DJOqa.NHe}saP3l?Ku9Jm3{[h" Like "#<@?*m*[?-W-O-bE-K]#Q*#?[]*N**?") = True
    
    End Sub
End Module
```

**Comment by Kr00l** [comment]
My proposal (2 cent) for that "quirk mode" setting would be a "Flags" value. (OR'ed Long's)
These "Flags" are saved in one setting and can be represented in a "reference" style list to check/uncheck the different quirks.
So it's tidied up and can be extended easily.

**Comment by mburns08109** [comment]
Is this possibly an artifact from VB(x) being developed referencing an older (buggy?) RegEx library version?

**Comment by WaynePhillipsEA** [comment]
@mburns08109 No, VBx doesn't use RegEx for this, it has a custom (buggy) parser.
